eras/mary/impl/testlib/Test/Cardano/Ledger/Mary/Arbitrary.hs

```diff
@@ -89,9 +98,88 @@ instance Crypto c => Arbitrary (PolicyID c) where
+--   2. All offsets (including those for AssetName and PolicyID) are relative to the whole
+--      of the representation (binary blob) rather than the start of their respective regions.
+--   3. If the offsets exceed their maxBounds, they will overflow.
+--   4. So, we need to ensure that at least the last of the offsets (AssetName offsets) do
+--      not exceed 65535.
+--   5. With `n` as the total number of assets, `p` the number of policy ids, the inequality to be satisfied is thus:
+--           8n -- Word64 asset quantities
+--        +  2n -- Word16 policy id offsets
+--        +  2n -- Word16 asset name offsets
+--        + 28p -- 28-byte policy ids
+--        + 32n -- 32-byte asset names (a maximum of 32 bytes)
+--        should be <= 65535, assuming the numer of policies to be maximal (i.e. equal to number of assets)
+--        65535 / 72 ~ 910.2 is the maximum number of triples to be safely generated
+--
+-- NOTE: There are some conditions due to which exceeding this number may not
+-- result in a guaranteed failure to compact without overflow, because, during compacting
+--   1. The asset names and policy ids are deduplicated
+--   2. Not all generated asset names are 32-bytes long
+-- But, exceeding this number does make the probability of causing overflow > 0.
+genMultiAsset :: forall c. Crypto c => Gen Integer -> Gen (MultiAsset c)
+genMultiAsset genAmount = do
+  MultiAsset ma <-
+    oneof
+      [ MultiAsset <$> genNonEmptyMap arbitrary (genNonEmptyMap arbitrary genAmount)
+      , multiAssetFromListBounded <$> listOf1 (genMultiAssetTriple $ fromIntegral <$> genAmount)
+      ]
+  if sum (length <$> ma) > 910
+    then scale (`div` 2) $ genMultiAsset genAmount
+    else pure $ MultiAsset ma
+
+-- | For tests, because `insertMultiAsset` called through `genMultiAsset` filters out zero values
+genMultiAssetZero :: Crypto c => Gen (MultiAsset c)
+genMultiAssetZero = MultiAsset <$> genNonEmptyMap arbitrary (genNonEmptyMap arbitrary (pure 0))
+
+-- For negative tests, we need a definite generator that causes overflow
+genMultiAssetToFail :: forall c. Crypto c => Gen (MultiAsset c)
+genMultiAssetToFail = do
+  (numPolicyId, numAssetName) <-
+    oneof
+      [ do
+          Positive numAsssetNames <- arbitrary
+          let minNumPolicyId = (65535 - 44 * numAsssetNames) `div` 28
+          numPolicyIds <- (minNumPolicyId +) . getPositive <$> arbitrary
+          pure (numPolicyIds, numAsssetNames)
+      , do
+          Positive numPolicyIds <- arbitrary
+          let minNumAssetNames = (65535 - 28 * numPolicyIds) `div` 44
+          numAssetNames <- (minNumAssetNames +) . getPositive <$> arbitrary
+          pure (numPolicyIds, numAssetNames)
+      ]
+  -- Here we generate separately a list of asset names and a list of policy ids and
+  -- randomly shuffle them into a MultiAsset
+
+  if numPolicyId > numAssetName -- Since we don't want to generate policies with empty assets
```

Comment:
Here is a simpler strategy that you can use:

Generate a list with `Int`s of length `numPolicyId` with positive numbers that sum up to `numAssetName`. Then you can use later to generate the nested structure. eg.
```
numPloicyId = 2
numAssetName = 5
```
Then one of lists that could be generated is:
```
lengths = [4, 1]
```

This would mean you can generate a multiasset with 2 policyIds where first one has 4 asset names and the second one has 1. This can be achived with two simple recursive functions.

Also. Do not use `resize` like this, this is not what it was designed for.

Reply:
> with positive numbers that sum up to numAssetName

I could not get to doing this today. However, I have updated the generator as per our previous discussion, and it results in sizes slightly above 65535 in most cases but also almost double that in some. Let me know what you think. I will be back to pick this up, if the current update does not   make the cut.